### ISSUE TITLE
Async bound functions in ES6 classes receive arguments of constructor.

### ISSUE THREAD

**Issue by outreach-devin-canterberry** [babel/babel#4723]
### Input Code

The following test suite describes the issue:

```js
import {expect} from 'chai';

class Demo {
  bad = async (x) => arguments[0];
  good = (x) => (async () => (x))();
}

describe('async bound handlers', function() {
  lazy('demo', () => new Demo('Scope leak!'));

  describe('#bad', function() {
    it('does bad things', async function() {
      const leak = await this.demo.bad(123);
      expect(leak).to.equal('Scope leak!');
    });
  });

  describe('#good', function() {
    it('does good things', async function() {
      const x = await this.demo.good(123);
      expect(x).to.equal(123);
    });
  });
});
```

### Babel Configuration (.bablerc, package.json, cli command)

Using [broccoli-babel-transpiler](https://github.com/babel/broccoli-babel-transpiler) with the following options:

```js
{
  nonStandard: true,
  moduleIds: false,
  modules: 'system',
  externalHelpers: true,
  stage: 0,
  // …other project-specific configuration (extensions, directories, etc)
}
```

### Expected Behavior

When defining a bound async function within an ES6 class, like so:

```javascript
class Demo {
  hello = async (name) => `Hello, ${name}!`;
}
```

I would expect `name` to be defined.

### Current Behavior

With the above example given for *expected behavior*, the value of `name` is bound to the first argument given to the constructor for `Demo`.

### Possible Solution

A workaround is to declare the bound function *without* the `async` keyword, and have it return a self-evaluating function *with* the `async` keyword containing the original logic. For example:

```javascript
class Demo {
  hello = (name) => (async () => `Hello, ${name}!`)();
}
```

A possible solution might be for Babel to do this automatically.

### Context

In practice, our React project passes bound functions as callbacks (e.g: `onClick`). Rather than dynamically binding these using `::` or `Function#bind`, which create new objects that violate shallow prop comparisons and trigger needless re-rendering (see [Performance Engineering with React](http://benchling.engineering/performance-engineering-with-react/#functionbindinlineanonymousfunctions)), we bind these using the pattern discussed here.

A fair number of our handler functions have asynchronous behavior, and we'd like to take advantage of the `async` keyword as we have with our unbound class functions. The obvious way to do this is to declare the assigned function to have the `async` keyword, which triggers the errant behavior.

In short, we want the `job` variable in both functions declared below to be consistent:

```javascript
class Demo {
  async performAsync(job) {/*…process the given job…*/}
  boundPerformAsync = async (job) => {/*…process the given job…*/}
}
```

`this.performAsync(job)` and `this.boundPerformAsync(job)` should be functionally equivalent.

**Comment by danharper** [comment]
Just to confirm, this is when using `arguments` inside a class property arrow function, right?

[REPL link](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Cstage-0&experimental=true&loose=false&spec=true&code=class%20Demo%20%7B%0A%20%20bad%20%3D%20async%20(name)%20%3D%3E%20arguments%5B0%5D%3B%0A%20%20good%20%3D%20async%20(name)%20%3D%3E%20name%3B%0A%7D%0A%0A(new%20Demo('x')).bad('y').then(console.log.bind(console))%3B%0A(new%20Demo('x')).good('y').then(console.log.bind(console))%3B&playground=false&stage=0)

```js
class Demo {
  bad = async (name) => arguments[0];
  good = async (name) => name;
}

(new Demo('x')).bad('y').then(console.log.bind(console)); // "x"
(new Demo('x')).good('y').then(console.log.bind(console)); // "y"
```

Arrow functions don't have `arguments` like normal functions do, so it comes from the outer scope. In this case, we're initializing in the constructor, so it comes from there. Will have to check the spec to be sure of the correct behaviour.

**Comment by danharper** [comment]
Closing as a duplicate of #2347, will be fixed with #4641.

**Comment by outreach-devin-canterberry** [comment]
I used arguments in the demo to demonstrate that the arguments are not the correct reference.
The code could be written like so to achieve the same result:

```javascript
class Demo {
  bad = async (name) => name;
}
new Demo('x').bad('y').then(console.log); // "x"
```

**Comment by danharper** [comment]
What version of Babel are you using, and which plugins/presets?

**Comment by danharper** [comment]
Ah just saw your edit :+1:
